rust/datafusion/src/datasource/datasource.rs

```diff
@@ -48,9 +66,19 @@ pub trait TableProvider {
         &self,
         projection: &Option<Vec<usize>>,
         batch_size: usize,
+        filters: &[Expr],
```

Comment:
That's an interesting idea! In my use case, the TableProvider impl interprets the filters synchronously after _asynchronously_ preparing the necessary data externally from DataFusion, but yeah, I could see that happening too.

If we didn't make this change, I suppose implementations could still use these filters within the returned `ExecutionPlan`'s `execute` method for now to enable async use cases?

Reply:
I think if filtering requires IO it could be done as part of the actual Execution. 

It seems like the main benefit of removing plan nodes during planning is to avoid potentially expensive I/O that could be ruled out with Metadata queries (as @returnString  describes is his usecase) 

Thus if the system has to go and do I/O anyways to figure out if it rule out a plan node, much of the benefit of pruning such nodes from the plan seem to be lost. In that case it seems better to me to do do such pruning as part of Execution.

Reply:
I agree that adding async requires quite a bit of propagation.